File: backend/app/routes/uploads.py
import json
from urllib.parse import urlparse

from flask import Blueprint, jsonify, request

from app.constants import ALLOWED_IMAGE_TYPES, MAX_INVOICE_SIZE_BYTES
from app.models.user import User
from app.services.image_upload_service import (
    ImageUploadError,
    complete_presigned_upload,
    create_presigned_upload,
    create_upload_presign,
    is_image_upload_configured,
    is_storage_configured,
    key_from_public_url,
    parse_content_length,
    parse_presign_fields,
)
from app.services.r2_service import build_object_key, build_unidentified_object_key
from app.utils.auth_decorators import jwt_required, staff_required

uploads_bp = Blueprint("uploads", __name__)

_ALLOWED_UPLOAD_HOST_SUFFIXES = (
    ".backblazeb2.com",
    ".r2.cloudflarestorage.com",
)


def _is_allowed_upload_url(url: str) -> bool:
    host = urlparse(url).netloc.lower()
    return any(host.endswith(suffix) for suffix in _ALLOWED_UPLOAD_HOST_SUFFIXES)


def _presign_error(exc: Exception, status: int = 500):
    if isinstance(exc, ImageUploadError):
        return jsonify({"error": str(exc)}), exc.status_code or status
    return jsonify({"error": f"Failed to generate upload URL: {exc}"}), status


@uploads_bp.route("/upload-url", methods=["POST"])
@jwt_required()
def proxy_upload_url():
    """Proxy presign requests to the image Worker (keeps API key server-side)."""
    if not is_image_upload_configured():
        return jsonify({"error": "Image upload worker is not configured"}), 503

    data = request.get_json(silent=True) or {}
    content_type = (data.get("content_type") or data.get("contentType") or "").strip().lower()
    prefix = (data.get("prefix") or "packages").strip() or "packages"
    try:
        content_length = parse_content_length(data)
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    if not content_type:
        return jsonify({"error": "content_type is required"}), 400

    try:
        return jsonify(
            create_presigned_upload(
                content_type=content_type,
                content_length=content_length,
                prefix=prefix,
            )
        )
    except ImageUploadError as exc:
        return _presign_error(exc)


@uploads_bp.route("/uploads/put", methods=["POST"])
@jwt_required()
def proxy_presigned_put():
    """Proxy file PUT to B2/R2 so browsers are not blocked by storage CORS."""
    if not is_storage_configured():
        return jsonify({"error": "File storage is not configured"}), 503

    upload_url = (request.form.get("upload_url") or "").strip()
    public_url = (request.form.get("public_url") or "").strip()
    object_key = (request.form.get("object_key") or "").strip()
    content_type = (request.form.get("content_type") or "application/octet-stream").strip().lower()

    if not upload_url:
        return jsonify({"error": "upload_url is required"}), 400
    if not _is_allowed_upload_url(upload_url):
        return jsonify({"error": "Invalid upload URL"}), 400

    file = request.files.get("file")
    if not file:
        return jsonify({"error": "file is required"}), 400

    file_bytes = file.read()
    if not file_bytes:
        return jsonify({"error": "file is empty"}), 400

    if len(file_bytes) > MAX_INVOICE_SIZE_BYTES:
        return jsonify({"error": "File exceeds maximum allowed size"}), 400

    try:
        upload_headers = json.loads(request.form.get("upload_headers") or "{}")
        if not isinstance(upload_headers, dict):
            upload_headers = {}
    except (TypeError, json.JSONDecodeError):
        upload_headers = {}

    try:
        complete_presigned_upload(
            upload_url=upload_url,
            file_bytes=file_bytes,
            content_type=content_type,
            upload_headers=upload_headers,
        )
    except ImageUploadError as exc:
        return _presign_error(exc, status=502)

    if not public_url and object_key:
        from app.services.r2_service import get_public_url

        public_url = get_public_url(object_key) or ""
    if not object_key and public_url:
        object_key = key_from_public_url(public_url)

    return jsonify({"public_url": public_url, "object_key": object_key})


@uploads_bp.route("/uploads/presign", methods=["POST"])
@staff_required()
def presign_upload():
    if not is_storage_configured():
        return jsonify({"error": "File storage is not configured"}), 503

    data = request.get_json(silent=True) or {}
    shipping_id = (data.get("shipping_id") or "").strip().upper()
    try:
        filename, content_type, content_length = parse_presign_fields(
            data,
            default_filename="photo.jpg",
            default_content_type="image/jpeg",
        )
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    if content_type not in ALLOWED_IMAGE_TYPES:
        return jsonify({"error": "Only JPEG, PNG, and WebP images are allowed"}), 400
    if not shipping_id:
        return jsonify({"error": "shipping_id is required"}), 400

    customer = User.query.filter_by(shipping_id=shipping_id).first()
    if not customer:
        return jsonify({"error": "Customer not found for that BOSS ID"}), 404

    try:
        payload = create_upload_presign(
            content_type=content_type,
            content_length=content_length,
            prefix="packages",
            r2_object_key=build_object_key(shipping_id, filename),
        )
    except ImageUploadError as exc:
        return _presign_error(exc)

    payload["shipping_id"] = shipping_id
    return jsonify(payload)


@uploads_bp.route("/uploads/presign-unidentified", methods=["POST"])
@staff_required()
def presign_unidentified_upload():
    if not is_storage_configured():
        return jsonify({"error": "File storage is not configured"}), 503

    data = request.get_json(silent=True) or {}
    try:
        filename, content_type, content_length = parse_presign_fields(
            data,
            default_filename="photo.jpg",
            default_content_type="image/jpeg",
        )
    except ValueError as exc:
        return jsonify({"error": str(exc)}), 400

    if content_type not in ALLOWED_IMAGE_TYPES:
        return jsonify({"error": "Only JPEG, PNG, and WebP images are allowed"}), 400

    try:
        return jsonify(
            create_upload_presign(
                content_type=content_type,
                content_length=content_length,
                prefix="packages",
                r2_object_key=build_unidentified_object_key(filename),
            )
        )
    except ImageUploadError as exc:
        return _presign_error(exc)

File: server/src/routes/uploads.ts
import { existsSync } from 'node:fs'
import { eq } from 'drizzle-orm'
import { Router } from 'express'
import multer from 'multer'
import { ALLOWED_IMAGE_TYPES, MAX_INVOICE_SIZE_BYTES } from '../constants.js'
import { db } from '../db/index.js'
import { users } from '../db/schema/index.js'
import { jwtRequired, requireAuth, warehouseRequired, type AuthRequest } from '../middleware/auth.js'
import {
  ImageUploadError,
  completePresignedUpload,
  createPresignedUpload,
  createUploadPresign,
  isAllowedLocalObjectKey,
  parsePresignFields,
  isLocalUploadEnabled,
  isStorageConfigured,
  keyFromPublicUrl,
  localFilePath,
  localPublicUrl,
  parseContentLength,
  saveLocalUpload,
} from '../services/imageUploadService.js'
import { RateLimitExceeded, assertUploadPresignAllowed } from '../services/rateLimitService.js'

export const uploadsRouter = Router()

const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: MAX_INVOICE_SIZE_BYTES } })

const ALLOWED_UPLOAD_HOST_SUFFIXES = ['.backblazeb2.com']

function isAllowedUploadUrl(url: string): boolean {
  if (url === 'local') return isLocalUploadEnabled()
  try {
    const host = new URL(url).hostname.toLowerCase()
    return ALLOWED_UPLOAD_HOST_SUFFIXES.some((suffix) => host.endsWith(suffix))
  } catch {
    return false
  }
}

function presignError(res: import('express').Response, exc: unknown, status = 500) {
  if (exc instanceof ImageUploadError) return res.status(exc.statusCode ?? status).json({ error: exc.message })
  return res.status(status).json({ error: `Failed to generate upload URL: ${exc}` })
}

uploadsRouter.post('/upload-url', jwtRequired, requireAuth, async (req: AuthRequest, res) => {
  try {
    assertUploadPresignAllowed(req.user!.id)
  } catch (exc) {
    if (exc instanceof RateLimitExceeded) return res.status(429).json({ error: exc.message })
    throw exc
  }

  if (!isStorageConfigured()) return res.status(503).json({ error: 'Image upload is not configured' })

  const data = req.body ?? {}
  const contentType = String(data.content_type ?? data.contentType ?? '')
    .trim()
    .toLowerCase()
  const prefix = String(data.prefix ?? 'packages').trim() || 'packages'

  try {
    const contentLength = parseContentLength(data)
    if (!contentType) return res.status(400).json({ error: 'content_type is required' })
    return res.json(await createPresignedUpload({ contentType, contentLength, prefix }))
  } catch (exc) {
    if (exc instanceof ImageUploadError) return presignError(res, exc)
    if (exc instanceof Error) return res.status(400).json({ error: exc.message })
    return presignError(res, exc)
  }
})

uploadsRouter.post('/uploads/put', jwtRequired, requireAuth, upload.single('file'), async (req: AuthRequest, res) => {
  if (!isStorageConfigured()) return res.status(503).json({ error: 'File storage is not configured' })

  const uploadUrl = String(req.body?.upload_url ?? '').trim()
  const publicUrl = String(req.body?.public_url ?? '').trim()
  let objectKey = String(req.body?.object_key ?? '').trim()
  const contentType = String(req.body?.content_type ?? 'application/octet-stream').trim().toLowerCase()

  if (!uploadUrl) return res.status(400).json({ error: 'upload_url is required' })
  if (!isAllowedUploadUrl(uploadUrl)) return res.status(400).json({ error: 'Invalid upload URL' })

  const file = req.file
  if (!file) return res.status(400).json({ error: 'file is required' })
  if (!file.buffer.length) return res.status(400).json({ error: 'file is empty' })
  if (file.buffer.length > MAX_INVOICE_SIZE_BYTES) {
    return res.status(400).json({ error: 'File exceeds maximum allowed size' })
  }

  if (uploadUrl === 'local') {
    if (!objectKey) objectKey = publicUrl ? keyFromPublicUrl(publicUrl) : ''
    if (!objectKey || !isAllowedLocalObjectKey(objectKey)) {
      return res.status(400).json({ error: 'Invalid object_key for local upload' })
    }
    try {
      await saveLocalUpload(objectKey, file.buffer)
    } catch (exc) {
      if (exc instanceof Error && exc.message.includes('Invalid')) {
        return res.status(400).json({ error: exc.message })
      }
      return presignError(res, exc)
    }
    return res.json({ public_url: localPublicUrl(objectKey), object_key: objectKey })
  }

  let uploadHeaders: Record<string, string> = {}
  try {
    const parsed = JSON.parse(String(req.body?.upload_headers ?? '{}'))
    if (parsed && typeof parsed === 'object') uploadHeaders = parsed as Record<string, string>
  } catch {
    uploadHeaders = {}
  }

  try {
    await completePresignedUpload({
      uploadUrl,
      fileBytes: file.buffer,
      contentType,
      uploadHeaders,
    })
  } catch (exc) {
    return presignError(res, exc, 502)
  }

  if (!objectKey && publicUrl) objectKey = keyFromPublicUrl(publicUrl)
  return res.json({ public_url: publicUrl, object_key: objectKey })
})

uploadsRouter.post('/uploads/presign', jwtRequired, warehouseRequired, async (req: AuthRequest, res) => {
  try {
    assertUploadPresignAllowed(req.user!.id)
  } catch (exc) {
    if (exc instanceof RateLimitExceeded) return res.status(429).json({ error: exc.message })
    throw exc
  }

  if (!isStorageConfigured()) return res.status(503).json({ error: 'File storage is not configured' })

  const data = req.body ?? {}
  const shippingId = String(data.shipping_id ?? '')
    .trim()
    .toUpperCase()
  try {
    const [, contentType, contentLength] = parsePresignFields(data, {
      defaultFilename: 'photo.jpg',
      defaultContentType: 'image/jpeg',
    })
    if (!ALLOWED_IMAGE_TYPES.has(contentType)) {
      return res.status(400).json({ error: 'Only JPEG, PNG, and WebP images are allowed' })
    }
    if (!shippingId) return res.status(400).json({ error: 'shipping_id is required' })

    const [customer] = await db.select().from(users).where(eq(users.shippingId, shippingId)).limit(1)
    if (!customer) return res.status(404).json({ error: 'Customer not found for that BOSS ID' })

    const payload = await createUploadPresign({ contentType, contentLength, prefix: 'packages' })
    return res.json({ ...payload, shipping_id: shippingId })
  } catch (exc) {
    if (exc instanceof ImageUploadError) return presignError(res, exc)
    if (exc instanceof Error) return res.status(400).json({ error: exc.message })
    return presignError(res, exc)
  }
})

uploadsRouter.post(
  '/uploads/presign-unidentified',
  jwtRequired,
  warehouseRequired,
  async (req: AuthRequest, res) => {
    try {
      assertUploadPresignAllowed(req.user!.id)
    } catch (exc) {
      if (exc instanceof RateLimitExceeded) return res.status(429).json({ error: exc.message })
      throw exc
    }

    if (!isStorageConfigured()) return res.status(503).json({ error: 'File storage is not configured' })

    const data = req.body ?? {}
    try {
      const [, contentType, contentLength] = parsePresignFields(data, {
        defaultFilename: 'photo.jpg',
        defaultContentType: 'image/jpeg',
      })
      if (!ALLOWED_IMAGE_TYPES.has(contentType)) {
        return res.status(400).json({ error: 'Only JPEG, PNG, and WebP images are allowed' })
      }
      return res.json(await createUploadPresign({ contentType, contentLength, prefix: 'packages' }))
    } catch (exc) {
      if (exc instanceof ImageUploadError) return presignError(res, exc)
      if (exc instanceof Error) return res.status(400).json({ error: exc.message })
      return presignError(res, exc)
    }
  },
)

uploadsRouter.get('/uploads/files/*', (req, res) => {
  if (!isLocalUploadEnabled()) return res.status(404).json({ error: 'Not found' })
  const objectKey = req.path.replace(/^\/uploads\/files\/?/, '')
  if (!isAllowedLocalObjectKey(objectKey)) return res.status(404).json({ error: 'Not found' })
  try {
    const path = localFilePath(objectKey)
    if (!existsSync(path)) return res.status(404).json({ error: 'Not found' })
    return res.sendFile(path)
  } catch {
    return res.status(404).json({ error: 'Not found' })
  }
})
